File: ConstructBinaryTree/ConstructBinaryTree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* reConstructBinaryTree(vector<int> pre,vector<int> vin) {
        if (pre.empty() || vin.empty())
            return NULL;
        int n = pre.size();
        // 前序遍历的第一个数字是根节点的值
        int root = pre[0];
        // 创建根节点
        TreeNode* t = new TreeNode(root);
        // 如果长度为1，直接返回根节点
        if (n==1) return t;
        // 找到root所在的位置，确定好前序和中序中左子树和右子树序列的范围
        int root_index = 0;
        for (int i=0;i<n;i++)
        {
            if (vin[i]==root)
            {
                root_index = i;
                break;
            }
        }
        // 左子树
        vector<int> left_pre,left_in, right_pre, right_in;
        for (int i=0;i<root_index;i++)
        {
            left_pre.push_back(pre[i+1]); // +1 是因为前序遍历的第一个节点是根节点
            left_in.push_back(vin[i]);
        }
        // 右子树
        for (int i=root_index+1;i<n;i++)
        {
            right_pre.push_back(pre[i]);
            right_in.push_back(vin[i]);
        }
        t->left = reConstructBinaryTree(left_pre,left_in);
        t->right = reConstructBinaryTree(right_pre,right_in);
        return t;
    }
};

File: ConstructBinaryTree/ConstructBinaryTree.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    # 返回构造的TreeNode根节点
    def reConstructBinaryTree(self, pre, tin):
        # write code here
        n = len(pre)
        if n==0:
            return None
        root = pre[0]
        t = TreeNode(root)
        if n==1:
            return t
        rooot_index = 0
        for i in range(n):
            if tin[i]==root:
                root_index = i
                break
        left_pre = []
        left_in = []
        right_pre = []
        right_in = []
        for i in range(root_index):
            left_pre.append(pre[i+1])
            left_in.append(tin[i])
        for i in range(root_index+1,n):
            right_pre.append(pre[i])
            right_in.append(tin[i])
        t.left = self.reConstructBinaryTree(left_pre,left_in)
        t.right = self.reConstructBinaryTree(right_pre,right_in)
        return t
### Solution 2:简洁实现
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    # 返回构造的TreeNode根节点
    def reConstructBinaryTree(self, pre, tin):
        # write code here
        if not pre or not tin:
            return None
        root = TreeNode(pre.pop(0))
        index = tin.index(root.val)
        root.left = self.reConstructBinaryTree(pre, tin[:index])
        root.right = self.reConstructBinaryTree(pre, tin[index + 1:])
        return root
